### ISSUE TITLE
OIDC login final redirect to callback URL fails with missing nonce

### ISSUE THREAD

**Issue by jmullo** [envoyproxy/envoy#36871]
*Title*: OIDC login final redirect to callback URI fails with missing nonce

*Description*:
Error: "state query param does not contain url or nonce"

URL in browser: https://our.site.com/cb?code=uuid-code&state=url%3Dhttps%3A%2F%2Four.site.com%2F

*Repro steps*:
OIDC login flow with Cognito as identity provider.

**Comment by jmullo** [comment]
Apparently there are two different ways Cognito can mess up here. It either drops the nonce from state and loses it completely. Or it drops the nonce from state and puts the nonce to callback URL as a separate query parameter.

For these kind of scenarios an option to disable the whole nonce validation in Envoy would be needed.

**Comment by steeni** [comment]
Root issue here is that cognito login page doesn't handle state value properly and the nonce doesn't stay in the state but gets raised directly to the url parameter and when coming back to callback nonce has disappeared from the state. In their documentation they do not support urlencoded json state values (and I guess this extends to any format, json just so common that it was mentioned). If we would ask them to fix it they probably just refer that section of their documentation. https://docs.aws.amazon.com/cognito/latest/developerguide/authorization-endpoint.html#get-authorize-request-parameters

Implementing the point 3 here https://github.com/envoyproxy/envoy/issues/36276 would probably help because then nonce would not be part of the state. Alternatively if the state value would be base64 urlencoded (base64 but with / and + changed to _ and - and no padding =) there would be less chance that it could break. Doing both would probably give most flexibility to handle with different oidc gateways. In the meantime I would recommend adding an option to disable nonce checking.

**Comment by nezdolik** [comment]
cc @mattklein123 @TAOXUY @tyxia @derekargueta

**Comment by zhaohuabing** [comment]
> Apparently there are two different ways Cognito can mess up here. It either drops the nonce from state and loses it completely. Or it drops the nonce from state and puts the nonce to callback URL as a separate query parameter.
> 
> For these kind of scenarios an option to disable the whole nonce validation in Envoy would be needed.
> 
> It's possible that Cognito doesn't like state being URL-encoded, and could work as base64-encoded(?)

`nonce` is supposed to be transparent to the OIDC provider, I didn't realize that some OIDC providers would inspect the `state` parameter and remove the `nonce` outside of it.

**Comment by zhaohuabing** [comment]
> Implementing the point 3 here https://github.com/envoyproxy/envoy/issues/36276 would probably help because then nonce would not be part of the state.

I'm not sure if this the right way to go. Nonce is an optional paramete in the standard. The oauth2 filter will need to know whether the provider supports the nonce parameter or not and supports both the nonce in the state and the standalone nonce, which complicate the implementation.

**Comment by steeni** [comment]
> > Implementing the point 3 here #36276 would probably help because then nonce would not be part of the state.
> 
> I'm not sure if this the right way to go. Nonce is an optional paramete in the standard. The oauth2 filter will need to know whether the provider supports the nonce parameter or not and supports both the nonce in the state and the standalone nonce, which complicate the implementation.

Yeah, I understand. I notice that there is already an implementation of base64url encode/decode here https://github.com/envoyproxy/envoy/blob/main/source/common/common/base64.cc#L245 Maybe it could be used instead of percent-encoding in filter.cc. In theory, that should fix the immediate issue.

**Comment by zhaohuabing** [comment]
@steeni rethinking this, your suggestion to move the nonce out of the `state` and into the `nonce` parameter aligns better with the standard. Even though it's optional, most IDPs already support it including oauth0, Google, AWS Cognito and Microsoft Entra, etc. 

cc @missBerg

**Comment by zhaohuabing** [comment]
> Implementing the point 3 here https://github.com/envoyproxy/envoy/issues/36276 would probably help because then nonce would not be part of the state.

Report back here: 

the `nonce` in the state paramter and the nonce parameter are two differenent things: the nonce in the [state parameter ](https://datatracker.ietf.org/doc/html/rfc6749#section-10.12)is used for csrf prevention and is applicable for both oauth2 and oidc. Please note that the OIDC spec defines a seperate [nonce parameter](https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest) ,which is specifically designed to prevent replay attacks and is unique to OIDC. More discussion about state and nonce parameters can be found in this comment: https://github.com/envoyproxy/envoy/pull/37050#issuecomment-2510418618


I've decided to raise #37473 to use base64url encode to fix this issue.

**Comment by zhaohuabing** [comment]
@jmullo  this issue can be closed. It's been fixed in https://github.com/envoyproxy/envoy/pull/37473.

**Comment by jmullo** [comment]
> @jmullo this issue can be closed. It's been fixed in #37473.

Waiting for a release to confirm.
